SteamKit2/SteamKit2/Steam/Discovery/SmartCMServerList.cs

```diff
@@ -274,59 +244,61 @@ private IPEndPoint GetNextServerCandidateInternal()
                     return null;
                 }
 
-                DebugWrite( $"Next server candidiate: {serverInfo.EndPoint}" );
-                return serverInfo.EndPoint;
+                DebugWrite( $"Next server candidiate: {serverInfo.Record.EndPoint} ({serverInfo.Record.ProtocolTypes})" );
+                return serverInfo.Record;
             }
         }
 
         /// <summary>
         /// Get the next server in the list.
         /// </summary>
+        /// <param name="supportedProtocolTypes">The minimum supported <see cref="ProtocolTypes"/> of the server to return.</param>
         /// <returns>An <see cref="System.Net.IPEndPoint"/>, or null if the list is empty.</returns>
-        public IPEndPoint GetNextServerCandidate()
+        public CMServerRecord GetNextServerCandidate( ProtocolTypes supportedProtocolTypes )
         {
             if ( !WaitForServersFetched() )
             {
                 return null;
             }
 
-            return GetNextServerCandidateInternal();
+            return GetNextServerCandidateInternal( supportedProtocolTypes );
         }
 
         /// <summary>
         /// Get the next server in the list.
         /// </summary>
+        /// <param name="supportedProtocolTypes">The minimum supported <see cref="ProtocolTypes"/> of the server to return.</param>
         /// <returns>An <see cref="System.Net.IPEndPoint"/>, or null if the list is empty.</returns>
-        public async Task<IPEndPoint> GetNextServerCandidateAsync()
+        public async Task<CMServerRecord> GetNextServerCandidateAsync( ProtocolTypes supportedProtocolTypes )
         {
             StartFetchingServers();
             await listTask.ConfigureAwait( false );
 
-            return GetNextServerCandidateInternal();
+            return GetNextServerCandidateInternal( supportedProtocolTypes );
         }
 
         /// <summary>
         /// Gets the <see cref="System.Net.IPEndPoint"/>s of all servers in the server list.
         /// </summary>
         /// <returns>An <see cref="T:System.Net.IPEndPoint[]"/> array contains the <see cref="System.Net.IPEndPoint"/>s of the servers in the list</returns>
-        public IPEndPoint[] GetAllEndPoints()
+        public CMServerRecord[] GetAllEndPoints()
```

Comment:
IReadOnlyList would be ideal, but ICollection gets optimised in LINQ.

I don’t see any issue with Array here, as it’s a copy of the internal data, not a mutable handle.